src/lib/adapters/waku/safe-waku.ts

```diff
@@ -0,0 +1,243 @@
+		let timeout = 1_000
+
+		const start = Date.now()
+
+		if (message.timestamp === this.lastSentTimestamp) {
+			message.timestamp++
+		}
+		this.lastSentTimestamp = message.timestamp
+
+		do {
+			try {
+				error = await sendMessage(this.lightNode, id, message)
+			} catch (e) {
+				error = e
+			} finally {
+				if (error) {
+					this.errors.numSendError++
+					this.log(`⁉️  Error: ${error}`)
+					this.log(`🕓 Waiting ${timeout} milliseconds...`)
+					await new Promise((r) => setTimeout(r, timeout))
+					if (timeout < 120_000) {
+						timeout += timeout
+					}
+				}
+			}
+		} while (error)
+
+		const elapsed = Date.now() - start
+
+		if (elapsed > 1000) {
+			this.log(`⏰ sending message took ${elapsed} milliseconds`)
+		}
+	}
+
+	private connectWaku() {
+		const waku = connectWaku({
+			onConnect: (connections) => {
+				this.log('✅ connected to waku', { connections })
+				this.safeResubscribe()
+
+				if (this.options?.onConnect) {
+					this.options?.onConnect(connections)
+				}
+			},
+			onDisconnect: () => {
+				this.log('❌ disconnected from waku')
+				this.errors.numDisconnect++
+
+				if (this.options?.onDisconnect) {
+					this.options.onDisconnect()
+				}
+			},
+		})
+		return waku
+	}
+
+	private async safeResubscribe() {
+		// eslint-disable-next-line no-constant-condition
+		while (true) {
+			try {
+				return await this.resubscribe()
+			} catch (e) {
+				console.error(`⁉️  ${e}`)
+			}
+		}
+	}
+
+	private async resubscribe() {
+		if (!this.lightNode) {
+			return
+		}
+
+		if (this.isSubscribing) {
+			return
+		}
+
+		this.isSubscribing = true
+
+		const chatIds = this.subscriptions.keys()
+		for (const subscription of this.subscriptions.values()) {
+			await subscription.unsubscribe()
+		}
+
+		const oldSubscriptions = this.subscriptions
+		this.subscriptions = new Map()
+
+		for (const chatId of chatIds) {
+			const callback = oldSubscriptions.get(chatId)?.callback
+			if (callback) {
+				await this.subscribe(chatId, undefined, callback)
+			}
+		}
+
+		this.isSubscribing = false
+	}
+
+	private async queueMessage(callback: Callback, chatMessage: Message, chatId: string) {
+		this.queuedMessages.push({
+			callback,
+			chatMessage,
+			chatId,
+		})
+
+		if (this.isHandlingMessage) {
+			return
+		}
+
+		this.isHandlingMessage = true
+
+		while (this.queuedMessages.length > 0) {
+			const queuedMessage = this.queuedMessages.shift()
+			if (queuedMessage) {
+				// deduplicate already seen messages
+				const message = queuedMessage.chatMessage
+				const lastMessage = this.lastMessages.get(chatId)
+				if (
+					lastMessage &&
+					lastMessage.timestamp === message.timestamp &&
+					lastMessage.type === message.type &&
+					lastMessage.fromAddress === message.fromAddress
+				) {
+					this.log('🙈 ignoring duplicate message', { message, lastMessage })
+					continue
+				}
+
+				this.lastMessages.set(chatId, message)
+
+				await callback(queuedMessage.chatMessage, queuedMessage.chatId)
```

Comment:
This could technically also throw and thus cancel `this.isHandlingMessage = false`.

Reply:
Yes, that's a good point, I will wrap this.